### PR TITLE
Add fine-tuning of entire model

### PR DESCRIPTION
**Pull request by W4ngatang**
Add ability to fine-tune entire model. PR includes:

- wrap checkpoint searching in a function in main to avoid code duplication during target task evaluation
- create a pre-target-task-training checkpoint if no pretraining
- during target task training, get all model parameters and save the resulting model checkpoint in a target-task-specific checkpoint
- during target task evaluation, load the correct target-task-specific fine-tuned model (and use MNLI for MNLI diagnostic)

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -212,11 +236,11 @@ def main(cl_arguments):
                                                             args.run_dir,
                                                             should_decrease)
         to_train = [(n, p) for n, p in model.named_parameters() if p.requires_grad]
-        best_epochs = trainer.train(pretrain_tasks, stop_metric,
-                                    args.batch_size, args.bpp_base,
-                                    args.weighting_method, args.scaling_method,
-                                    to_train, opt_params, schd_params,
-                                    args.shared_optimizer, args.load_model, phase="main")
+        _ = trainer.train(pretrain_tasks, stop_metric,
```
Comment (W4ngatang) [inline main.py]: not related to PR, but this wasn't used. I can revert if necessary.

File: main.py [hunk 2]
```diff
@@ -236,104 +260,147 @@ def main(cl_arguments):
...
     if args.do_target_task_training:
-        # might be empty if no elmo. scalar_mix_0 should always be pretrain scalars
-        elmo_scalars = [(n, p) for n, p in model.named_parameters() if
-                        "scalar_mix" in n and "scalar_mix_0" not in n]
-        # fails when sep_embs_for_skip is 0 and elmo_scalars has nonzero length
-        assert_for_log(not elmo_scalars or args.sep_embs_for_skip,
-                       "Error: ELMo scalars loaded and will be updated in do_target_task_training but "
-                       "they should not be updated! Check sep_embs_for_skip flag or make an issue.")
+        if args.transfer_paradigm == "frozen":
+            # might be empty if elmo = 0. scalar_mix_0 should always be pretrain scalars
```
Comment (sleepinyourhat) [inline main.py]: Rephrase to make it clearer that it should be empty if sep_embs_for_skip is 0.

File: main.py [hunk 3]
```diff
@@ -236,104 +260,147 @@ def main(cl_arguments):
...
             if task.name == 'mnli-diagnostic':
                 continue
-            pred_module = getattr(model, "%s_mdl" % task.name)
-            to_train = elmo_scalars + [(n, p)
-                                       for n, p in pred_module.named_parameters() if p.requires_grad]
+
+            if args.transfer_paradigm == "finetune":
+                # Train both the task specific models as well as sentence encoder.
+                to_train = [(n, p) for n, p in model.named_parameters() if p.requires_grad]
+            elif args.transfer_paradigm == "frozen":
```
Comment (sleepinyourhat) [inline main.py]: else?

File: main.py (hunk 2 above)
Comment (W4ngatang) [reply main.py]: It can be non-empty even if `sep_embs_for_skip = 0` since there are still target-task specific mixing weights.

File: main.py [hunk 4]
```diff
@@ -236,104 +263,148 @@ def main(cl_arguments):
...
+                                       shared_optimizer=args.shared_optimizer,
+                                       load_model=False, phase="eval")
 
             # Now that we've trained a model, revert to the normal checkpoint logic for this task.
             task_names_to_avoid_loading.remove(task.name)
 
             # The best checkpoint will accumulate the best parameters for each task.
             # This logic looks strange. We think it works.
             best_epoch = best_epoch[task.name]
+            # TODO(Yada): Move logic for checkpointing finetuned vs frozen pretrained tasks
```
Comment (W4ngatang) [inline main.py]: @pruksmhc ?

File: main.py (hunk 2 above)
Comment (W4ngatang) [reply main.py]: I'm confused. If `sep_embs_for_skip = 0`, isn't it possible to have elmo scalars to train (task-specific layer mixing weights)?

File: main.py [hunk 5]
```diff
@@ -82,6 +82,30 @@ def _kill_tb_child():
     atexit.register(_kill_tb_child)
 
 
+def get_best_checkpoint_path(run_dir):
```
[the remark this answers is not in the document]
Comment (pruksmhc) [reply main.py]: Yep, there's a todo down to move this checkpointing logic to trainer.py where the rest of checkpointing lives- even better, it should live in its own file

File: main.py (hunk 4 above)
Comment (pruksmhc) [reply main.py]: Move to trainer.py or its own file (see comment above w/ Sam)

File: main.py [hunk 6]
```diff
@@ -20,13 +20,15 @@
...
-
 from src.utils.utils import assert_for_log, maybe_make_dir, load_model_state, check_arg_name
 from src.preprocess import build_tasks
 from src.models import build_model
 from src.trainer import build_trainer, build_trainer_params
 from src import evaluate
 
+# Global notification handler, can be accessed outside main() during exception
+# handling.
+EMAIL_NOTIFIER = None
```
Comment (W4ngatang) [inline main.py]: just moving this to before all the function definitions

File: main.py (hunk 2 above)
Comment (W4ngatang) [reply main.py]: Oh, there are no other ELMo scalars if no skip connection b/c there are no task-specific weights in that case?

Comment (sleepinyourhat) [reply main.py]: Er, I probably just misunderstood this. Nvm.

**Pull request merged**